src/library/scala/collection/Iterator.scala

```diff
@@ -1189,6 +1223,80 @@ object Iterator extends IterableFactory[Iterator] {
       } else Iterator.empty.next()
     }
   }
+
+  private final class TakeRightIterator[A](private[this] var underlying: Iterator[A], buflen: Int) extends AbstractIterator[A] {
+    private[this] var len: Int = -1
+    private[this] var pos: Int = 0
+    private[this] var buf: Array[AnyRef] = _
+    def init(): Unit = if(buf eq null) {
+      buf = new Array[AnyRef](buflen)
+      len = 0
+      while(underlying.hasNext) {
+        buf(pos) = underlying.next().asInstanceOf[AnyRef]
+        pos = (pos + 1) % buflen
```

Comment:
It's faster to do a jump than modulus arithmetic, i.e. something like `pos += 1; if (pos == buflen) pos = 0` (assuming `buflen` is sensible, which it had better be if an array was allocated using it).